pkg/acs/calacs/acsccd/doccd.c

```diff
@@ -143,278 +171,582 @@ int DoCCD (ACSInfo *acs_info) {
 
         if (PutKeyStr(&x[i].sci.hdr, "BUNIT", "ELECTRONS", "")) {
-            freeSingleGroup(&x[i]);
+            freeOnExit (&ptrReg);
             return (status);
         }
         if (PutKeyStr(&x[i].err.hdr, "BUNIT", "ELECTRONS", "")) {
-            freeSingleGroup(&x[i]);
+            freeOnExit (&ptrReg);
             return (status);
         }
-    }
+    }}
     /************************************************************************/
 
     /************************************************************************/
     /* Subtract bias level determined from overscan. */
-    BlevMsg(&acs[0], 1);
+    BlevMsg(&acs[primaryIdx], nextver);
 
-    for (i = 0; i < acs_info->nimsets; i++) {
-        blevcorr[i] = PERFORM;
+    int * blevcorr = NULL;
+    blevcorr = (int *) malloc(acs_info->nimsets * sizeof(int));
+    if (!blevcorr) {
+       freeOnExit (&ptrReg);
+       return (OUT_OF_MEMORY);
     }
+    addPtr (&ptrReg, blevcorr, free);
 
-    /* default blev case, do original bias level subtraction */
-    if (acs_info->blevcorr == PERFORM &&
-        ((acs_info->expstart < SM4MJD && acs_info->detector == WFC_CCD_DETECTOR) ||
-         (acs_info->detector == WFC_CCD_DETECTOR && acs_info->subarray == YES) ||
-         (acs_info->detector != WFC_CCD_DETECTOR))) {
-        for (i = 0; i < acs_info->nimsets; i++) {
-            /* This is set based on results of FindOver */
-            done = overscan[i];
-
-            if (doBlev(&acs[i], &x[i], acs[i].chip, &meanblev, &done,
-                       &driftcorr))
-                return (status);
-
-            if (done) {
-                trlmessage("Bias level from overscan has been subtracted;");
+    {unsigned int i;
+    for (i = 0; i < acs_info->nimsets; i++) {
+        blevcorr[i] = PERFORM;
+    }}
+
+    /* The logic here has been re-written (1) to accommodate the new subarrays
+       (Ref: ISR ACS 2017-03) which can be bias shift corrected ("*-2K" data only at 
+       this time), and (2) the head of the ACS team requested the logic be clarified.  
+       This code is verbose and potentially harder to maintain, but the logic is clearer
+       for the scientist reading the code.
+
+       NOTE: The variable "done" is overloaded.  When done = overscan[i], this means
+       a column was matched in the OSCNTAB for the image and there is physical overscan
+       in the image.  Variable "done" is updated in doblev() to indicate the bias has
+       been determined from the overscan region rather than using a default bias obtained
+       from CCDTAB.
+    */
 
-                if (PutKeyFlt(&x[i].sci.hdr, "MEANBLEV", meanblev,
-                              "mean of bias levels subtracted in electrons")) {
+    int done = NO;       
+    int driftcorr = NO;     /* true means bias level was corrected for drift */
+    Bool isSupportedSubAperture = NO;
+    float meanblev = 0.0;   /* mean value of overscan bias (for history) */
+    if (acs_info->blevcorr == PERFORM) {
```

Comment:
I would pull this entire block out into a sep func. This would help clear up the crowded logic. Same prob goes for all the other steps.